### PR TITLE
Feature/key listener

### PR DESCRIPTION
**Pull request by bogdzn**
- basic key listener
- unit tests
- improved documentation
- refactored exec loop
- refactored parser for key listener
- colored logs !
- reworked file structure

# coverage: 76%

### CODE REVIEW

File: app/include/api_commands.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <asm-generic/int-ll64.h>
+#include <linux/input-event-codes.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include "parser.h"
+#include "utils.h"
+
+/// this file should only be refecterenced in get_instructions_queue.c
```
Comment (darksailormoonlight) [inline app/include/api_commands.h:20]: referenced *

File: app/include/api_commands.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * \struct authorized_t
+ * \brief a function authorized by the API.
+ *
+ * \param command the command you want to send to your drone
+ * \param code code defined in <linux/input-event-codes.h>
+ */
+typedef struct authorized_s
+{
+    char *command;
+    char code;
```
Comment (darksailormoonlight) [inline app/include/api_commands.h:32]: keycode

File: app/include/parser.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
+/**
+ * \file parser.h
+ * \brief Handles initial parsing.
+ * \author Adina C.
```
Comment (darksailormoonlight) [inline app/include/parser.h:4]: it isn't mine anymore ...

File: app/include/parser.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+ * \brief user-defined settings
+ *
+ * \param max_timeout defines the number of the time a command will be retied if it fails.
+ * \param sleep_time defines the amout of time we wait before sending another command.
+ * \param fake_socket allows program to pretend it sends information to a drone.
+ * \param filepath path of the file full of commands.
+ */
+typedef struct input_s settings_t;
+struct input_s {
+    int max_timeout;
```
Comment (darksailormoonlight) [inline app/include/parser.h:37]: max retries

File: app/include/parser.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    unsigned int sleep_time;
```
Comment (darksailormoonlight) [inline app/include/parser.h:38]: timeout

File: app/src/exec/exec_loop.c [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+int exec_loop(socket_t ryze, settings_t settings, char **cmds)
+{
+    int status = 0;
+    char *drone_response = NULL;
+
+
+    for (int i = 0; cmds[i] && status == 0; i++) {
+        for (int timeout = 0; timeout < settings.max_timeout; timeout++) {
```
Comment (darksailormoonlight) [inline app/src/exec/exec_loop.c:33]: retry

File: app/src/exec/exec_loop.c [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            if (is_drone_ok(drone_response))
+                break;
+            __log(WARNING, "command %[s] failed...Retrying [%i/%i]\n",
+            cmds[i], timeout, settings.max_timeout);
+            if (timeout + 1 == settings.max_timeout) {
+                __log(ERROR, "command [%s] timed out, exiting...");
+                return -1;
+            }
+        }
+        sleep(settings.sleep_time);
```
Comment (darksailormoonlight) [inline app/src/exec/exec_loop.c:45]: wait_cmd or just wait

File: app/src/exec/key_listener.c [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    // resetting keyboard mode
+    if (is_raw_mode) {
+        tcsetattr(STDIN_FILENO, TCSAFLUSH, &tty_attr_old);
+        ioctl(STDIN_FILENO, KDSKBMODE, old_kb_mode);
+        is_raw_mode = false;
+        return 0;
+    } else if (!isatty(STDIN_FILENO)) {
+        __log(ERROR, "not in a tty.\n");
+        return -1;
+    }
```
Comment (darksailormoonlight) [inline app/src/exec/key_listener.c:43]: faire fonction à part

File: app/src/exec/key_listener.c [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+{
+    int status = 0;
+    char **commands = NULL;
+    char buffer = 0;
+
+    if (set_keyboard_mode() == -1) {
+        __log(ERROR, "Couldn't set terminal in raw mode.\n");
+        return 0;
+    }
+    send_command(ryze, "command", settings);
```
Comment (darksailormoonlight) [inline app/src/exec/key_listener.c:74]: commentaire initialiser la communication

File: app/src/exec/key_listener.c [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        set_keyboard_mode();
+        return 0;
+    }
+    while (status != 1) {
+        do {
+            usleep(75000);
+            status = read(STDIN_FILENO, &buffer, 1);
+        } while (status < 0);
+        if (buffer == 'c') {
+            __log(WARNING, "c has been triggered. Exiting.\n");
```
Comment (darksailormoonlight) [inline app/src/exec/key_listener.c:87]: if buffer == 'C' because of active shift

**Review by darksailormoonlight — APPROVED** [review]
do the changes !!!!

File: app/include/api_commands.h (hunk 1 above)
Comment (bogdzn) [reply app/include/api_commands.h:20]: done

File: app/include/api_commands.h (hunk 2 above)
Comment (bogdzn) [reply app/include/api_commands.h:32]: done

File: app/include/parser.h (hunk 3 above)
Comment (bogdzn) [reply app/include/parser.h:4]: done

File: app/include/parser.h (hunk 4 above)
Comment (bogdzn) [reply app/include/parser.h:37]: done

File: app/include/parser.h (hunk 5 above)
Comment (bogdzn) [reply app/include/parser.h:38]: done

File: app/src/exec/exec_loop.c (hunk 6 above)
Comment (bogdzn) [reply app/src/exec/exec_loop.c:33]: done

File: app/src/exec/key_listener.c (hunk 8 above)
Comment (bogdzn) [reply app/src/exec/key_listener.c:43]: done

File: app/src/exec/key_listener.c (hunk 10 above)
Comment (bogdzn) [reply app/src/exec/key_listener.c:87]: done

File: app/src/exec/key_listener.c (hunk 9 above)
Comment (bogdzn) [reply app/src/exec/key_listener.c:74]: done

**Pull request merged**